File: tests/lambda1.exs

defmodule Lambda1 do

  def var(id),       do: {:Var, id}
  def lam(id, body), do: {:Lam, id, body}
  def apl(lhs, rhs), do: {:Apl, lhs, rhs}

  def subst(var, val, term) do
    case term do
      {:Lam, ^var, _} -> term
      {:Lam, id, body} -> lam(id, subst(var, val, body))

      {:Var, ^var} -> val
      {:Var, _} -> term

      {:Apl, lhs, rhs} ->
        apl(subst(var, val, lhs), subst(var, val, rhs))
    end
  end

  def eval(term) do
    case term do
      {:Lam, _, _} -> term
      {:Apl, lhs, rhs} ->
        {:Lam, arg, body} = eval(lhs)
        rhs = eval(rhs)
        eval(subst(arg, rhs, body))
    end
  end

  def succ, do: lam(1, lam(2, lam(3,
    apl(var(2), apl( apl(var(1), var(2)), var(3) ))
  )))

  def add, do: lam(1, lam(2,
    apl( apl(var(1), succ() ), var(2) )
  ))

  def mul, do: lam(1, lam(2, lam(3,
    apl( var(1), apl(var(2), var(3)) )
  )))

  def pred, do: lam(1, lam(2, lam(3,
    apl(
        apl(
            apl( var(1), lam(4, lam(5, apl(var(5), apl(var(4), var(2))))) ),
            lam(6, var(3))
        ),
        lam(6, var(6))
    )
  )))

  def sub, do: lam(1, lam(2,
    apl( apl(var(2), pred()), var(1) )
  ))

  def church(0), do: lam(0, lam(1, var(1)))
  def church(n) do
    apl(succ(), church(n - 1))
  end

  def identity, do:
    lam(0, apl( apl(var(0), succ()), church(0) ))



	def get_input() do
		IO.binread(:line)
			|> String.trim()
			|> String.to_integer
	end

  def main() do
    n = get_input()

    one = apl( apl(sub(), church(n)), church(n - 1) )

    IO.puts(
      inspect( eval(apl(identity(), one)) )
    )
  end

end


Lambda1.main()

File: tests/lambda1.py

import sys
from collections import namedtuple


Var = namedtuple("Var", "id")

Lam = namedtuple("Lam", "arg body")

Apl = namedtuple("Apl", "lhs rhs")


def subst(var, value, term):
    if isinstance(term, Var):
        if term.id == var:
            return value
        else:
            return term

    elif isinstance(term, Lam):
        if term.arg == var:
            return term
        else:
            return Lam(term.arg, subst(var, value, term.body))
    
    elif isinstance(term, Apl):
        return Apl(
            subst(var, value, term.lhs),
            subst(var, value, term.rhs),
        )


def eval(term):
    if isinstance(term, Lam):
        return term    
    elif isinstance(term, Apl):
        lhs = eval(term.lhs)
        rhs = eval(term.rhs)
        assert isinstance(lhs, Lam)
        return eval(subst(lhs.arg, rhs, lhs.body))


succ = Lam(1, Lam(2, Lam(3, 
    Apl(Var(2), Apl(Apl(Var(1), Var(2)), Var(3)))
)))

add = Lam(1, Lam(2,
    Apl(Apl(Var(1), succ), Var(2))
))

mul = Lam(1, Lam(2, Lam(3,
    Apl(Var(1), Apl(Var(2), Var(3)))
)))

pred = Lam(1, Lam(2, Lam(3,
    Apl(
        Apl(
            Apl( Var(1), Lam(4, Lam(5, Apl(Var(5), Apl(Var(4), Var(2))))) ),
            Lam(6, Var(3))
        ),
        Lam(6, Var(6))
    )
)))

sub = Lam(1, Lam(2, 
    Apl(Apl(Var(2), pred), Var(1))
))

def church(n):
    term = Lam(0, Lam(1, Var(1)))
    for _ in range(n):
        term = Apl(succ, term)
    return term

identity = Lam(0,
    Apl(Apl(Var(0), succ), church(0))
)

if __name__ == "__main__":
    n = int(input())
    one = Apl(Apl(sub, church(n)), church(n - 1))
    sys.setrecursionlimit(n * 10)
    print(eval(Apl(identity, one)))